### PR TITLE
simplify unqualified col name aliases

### LINKED ISSUE #3722: Use simpler schema names for expressions
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

We currently use overly verbose schema names for expressions. Here is an example comparing scalar function names between Postgres and DataFusion.

## Postgres

```
postgres=# select concat(a, 'utf8', true, false, null, 1, 0.2) from t;select concat(a, 'utf8', true, false, null, 1, 0.2) from t;
   concat    
-------------
...
```

## DataFusion

```
❯ select concat(a, 'utf8', true, false, null, 1, 0.2) from t;
+----------------------------------------------------------------------------------+
| concat(t.a,Utf8("utf8"),Boolean(true),Boolean(false),NULL,Int64(1),Float64(0.2)) |
+----------------------------------------------------------------------------------+
...
```

**Describe the solution you'd like**
Use simple names in schema - use the function name in this example and do not add the parameters.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #3722 .

 # Rationale for this change

Use simple names in schema - use the function name in this example and do not add the parameters.

# What changes are included in this PR?

Use simple names in schema - use the function name in this example and do not add the parameters.

# Are there any user-facing changes?

The unqualified column name will be changed

### CODE REVIEW

**Comment by comphead** [comment]
@andygrove please check this PR

File: datafusion/core/tests/sql/select.rs [hunk 1]
```diff
@@ -505,14 +505,14 @@ async fn use_between_expression_in_select_query() -> Result<()> {
...
     let expected = vec![
-        "+-------------------------------------------------------------+",
-        "| abs(test.c1) BETWEEN Int64(0) AND log(test.c1 * Int64(100)) |",
-        "+-------------------------------------------------------------+",
-        "| true                                                        |",
-        "| true                                                        |",
-        "| false                                                       |",
-        "| false                                                       |",
-        "+-------------------------------------------------------------+",
+        "+------------------------------+",
```
Comment (alamb) [inline datafusion/core/tests/sql/select.rs:508]: The plan changes sure look nicer 😍

**Review by alamb — DISMISSED** [review]
I think it looks great -- thank you @comphead

**Comment by Dandandan** [comment]
Not sure I'm a fan. What happens with the alias if you use a function on more than one column?
E.g. `select trim(a), trim(b) from table`?

**Review by Dandandan — CHANGES_REQUESTED** [review]
What happens when using the same function more than once?

**Comment by alamb** [comment]
Here is what postgres does:

```
postgres=# select * from foo;
 val | time 
-----+------
 A   |  1.0
 B   |  2.0
 C   |  3.0
(3 rows)

postgres=# select trim(val), trim(val) from foo;
 btrim | btrim 
-------+-------
 A     | A
 B     | B
 C     | C
(3 rows)

postgres=# 
```

Here is what datafusion-cli does from master:

```
DataFusion CLI v13.0.0
❯ create table foo as select column1 as val, column2 as time from (values ('A', 1.0), ('B', 2.0), ('C', 3.0)) as sq;

0 rows in set. Query took 0.023 seconds.
❯ select trim(val), trim(val) from foo;
Plan("Projections require unique expression names but the expression \"trim(foo.val)\" at position 0 and \"trim(foo.val)\" at position 1 have the same name. Consider aliasing (\"AS\") one of them.")
```

**Comment by comphead** [comment]
> What happens when using the same function more than once?

@Dandandan thanks for you comment, the engine historically fails if the alias are the same,  like if you run
`let sql = "SELECT COALESCE(NULL, 'test'), COALESCE(NULL, 'test')";`
it will fail even now.

supposingly we need a separate ticket to try to do postgres behaviuor, or add some number like 
`coalesce_0, coalesce_1`. I'm happy to do this if needed

**Comment by Dandandan** [comment]
Ok, but the current branch now also fails with
`select trim(a), trim(b) from table`? As the two are now giving the same alias. This is my main worry to merge this. As suffix that @comphead suggests seems a solution.

**Comment by comphead** [comment]
> Ok, but the current branch now also fails with `select trim(a), trim(b) from table`? As the two are now giving the same alias. This is my main worry to merge this. As suffix that @comphead suggests seems a solution.

This is correct, as we shortened the column name the collision will happen more often. But this gives us more motivation to make unique readable aliases.

Because currently user may not be happy to refer to column from outer sql to nested subquery which contains not friendly alias like `btrim(test.a,Utf8(\"ab\"))`

**Comment by comphead** [comment]
@alamb @Dandandan should I put unique col aliasing like `trim_0, trim_1` as part of this PR, or create a new one?

**Comment by alamb** [comment]
> @alamb @Dandandan should I put unique col aliasing like trim_0, trim_1 as part of this PR, or create a new one?

I am not sure -- perhaps @andygrove  can weigh in here as he filed https://github.com/apache/arrow-datafusion/issues/3722

**Comment by comphead** [comment]
> > @alamb @Dandandan should I put unique col aliasing like trim_0, trim_1 as part of this PR, or create a new one?
> 
> I am not sure -- perhaps @andygrove can weigh in here as he filed #3722

@andygrove please help here until the code rots

**Comment by andygrove** [comment]
I will catch up on this and other pending PRs tomorrow.

**Comment by alamb** [comment]
Marking this as a draft as it is waiting on feedback and not ready for review or merge

**Comment by andygrove** [comment]
Postgres allows duplicate column names in a schema but then throws an error if you try and select one of those columns. We currently fail on creating a schema with duplicate columns.

I guess the first question is, do we want to be more like Postgres here?

**Comment by comphead** [comment]
> Postgres allows duplicate column names in a schema but then throws an error if you try and select one of those columns. We currently fail on creating a schema with duplicate columns.
> 
> I guess the first question is, do we want to be more like Postgres here?

Thanks @andygrove for the feedback.

Most likely we should avoid current naming for not qualified columns as its not user friendly and hard to use in outer subqueries.

And perhaps its a bigger topic to discuss naming standard, to be like a Postgres or something else, but we can create a discussion ticket on this first, to get all options from contributors.

If you ask me, I like how its done in Trino

`
select t.*, 'col_again' from (
	select 'col', 2 a, count(1)
) t
`

```
_col0|a|_col2|_col3    |
-----+-+-----+---------+
col  |2|    1|col_again|
```

**Comment by alamb** [comment]
Closing as this PR is over a year old. Please feel free to reopen it / rebase it if you plan to keep working on it

**Pull request closed without merging**